### ISSUE TITLE
__to_jax_array__  removed from version 0.10.2 to 0.10.3

### ISSUE THREAD

**Issue by liukidar** [patrick-kidger/equinox#488]
Hello,

I'm building a library on top of equinox (we chatted a little in the past about that if you remember) and I noticed that from version 0.10.2 to 0.10.3 you removed the `_to_jax_array` method in _conv.py. In the comments above those function you refer to JAX issue #10065 which, to my understanding is still open. The code I use wraps every jax.array in a Parameter class so the use of __jax_array__ is fundamental and, since 0.10.3, convolutions don't work anymore. By reading #10065 the problem is that my Parameter is both a pytree (that gets flatten to its value basically) and a __jax_array__able object.

My question is, did you remove `_to_jax_array` because of the issues rised in 10065? Did you find another solution to it? Or is it not supposed to be necessary anymore and what I am experiencing is a JAX bug?

As a solution I can have Parameter not be a pytree that gets converted to a FlattenableParameter when flattening the model it is contained into, but it seems like a overcomplicated solution. What do you think?

Thank you :)

**Comment by patrick-kidger** [comment]
Right! Quite a lot to unpack here.

First, why this was there in the first place: `_to_jax_array` was there specifically to make `eqx.experimental.SpectralNorm` work when applied to a convolutional layer. The usage was `eqx.tree_at(lambda c: c.weight, conv, eqx.experimental.SpectralNorm(conv.weight))`, so that it actually replaced the weight matrix. In order to work around JAX bug 10065, we needed to call `__jax_array__` manually.

Second, why it was removed: in Equinox v0.10.3, I figured out a new (much more performant, much less hacky) way to implement stateful layers like `SpectralNorm`. For this reason, the layer got rewritten as `eqx.nn.SpectralNorm`, now with a whole-new API, and `__jax_array__` was no longer needed. Correspondingly, I also deleted the hack that was `_to_jax_array`.

Third, the nature of the JAX bug itself: it's actually nothing to do with being a pytree or not, really. It's just that many operations in the JAX API don't actually call `__jax_array__`. (Moreover, it is planned to remove this API from JAX; it's since been decided that this is a bad idea.)

Fourth, and most importantly: what can you do to make your code work? (!) It turns out there's a fair simply fix: build up your pytree of parameters, and then tree-map over them to resolve them in to the actual arrays that are used at runtime. [See this guide for details.](https://docs.kidger.site/equinox/tricks/#custom-per-parameter-behaviour) Moreover this has the advantage that it will work for any layer at all, without needing special-case support a la `_to_jax_array`.

**Comment by liukidar** [comment]
I see, makes sense. I ended up creating a function that "jaxify" any model by extracting all the parameters and storing them in a "state" object while substituting in the raw vectors as you suggested.

You mentioned "much more performant", what do you mean exactly? Is it compilation or execution performance? From my understanding all this playing around and transforming pytrees doesn't affect the final execution graph, but maybe I'm wrong. As performance and efficiency is the first concern of the library I'm working on, I'd be very grateful to hear any advice related to it (or caveats I should be aware of).

Thank you for your help!

**Comment by patrick-kidger** [comment]
The performance issue that was fixed was specifically to do with stateful layers (not `__jax_array__`). These saved and loaded their state using callbacks, which introduced spurious copies.

So you're completely correct that our playing around with pytrees doesn't affect the final execution graph, and that changes of that nature don't affect runtime performance. I think you should be in the clear here!

**Comment by liukidar** [comment]
Great! Good to know :)
